plugins/omelasticsearch/omelasticsearch.c

```diff
@@ -1206,50 +1274,93 @@ curlResult(void *ptr, size_t size, size_t nmemb, void *userdata)
+	}
+
+	r = es_addBuf(&auth, uid, strlen(uid));
+	if(r == 0) r = es_addChar(&auth, ':');
+	if(r == 0 && pwd != NULL) r = es_addBuf(&auth, pwd, strlen(pwd));
+	if(r == 0) *authBuf = (uchar*) es_str2cstr(auth, NULL);
+
+	if (r != 0 || *authBuf == NULL) {
+		errmsg.LogError(0, RS_RET_ERR, "omelasticsearch: failed to build auth header\n");
+		ABORT_FINALIZE(RS_RET_ERR);
 	}
 
-	header = curl_slist_append(NULL, "Content-Type: text/json; charset=utf-8");
+finalize_it:
+	if (auth != NULL)
+		es_deleteStr(auth);
+	RETiRet;
+}
+
+static void
+curlCheckConnSetup(CURL *handle, HEADER *header, long timeout, sbool allowUnsignedCerts) 
+{
 	curl_easy_setopt(handle, CURLOPT_HTTPHEADER, header);
+	curl_easy_setopt(handle, CURLOPT_NOBODY, TRUE);
+	curl_easy_setopt(handle, CURLOPT_TIMEOUT_MS, timeout);
+	curl_easy_setopt(handle, CURLOPT_NOSIGNAL, TRUE);
+
+	if(allowUnsignedCerts)
+		curl_easy_setopt(handle, CURLOPT_SSL_VERIFYPEER, FALSE);
+
+	/* Only enable for debugging 
+	curl_easy_setopt(curl, CURLOPT_VERBOSE, TRUE); */
+}
 
+static void
+curlPostSetup(CURL *handle, HEADER *header, uchar* authBuf)
+{
+	curl_easy_setopt(handle, CURLOPT_HTTPHEADER, header);
 	curl_easy_setopt(handle, CURLOPT_WRITEFUNCTION, curlResult);
 	curl_easy_setopt(handle, CURLOPT_POST, 1);
 	curl_easy_setopt(handle, CURLOPT_NOSIGNAL, TRUE);
 
-	pWrkrData->curlHandle = handle;
-	pWrkrData->postHeader = header;
+	if(authBuf != NULL) {
+		curl_easy_setopt(handle, CURLOPT_USERPWD, authBuf);
+		curl_easy_setopt(handle, CURLOPT_PROXYAUTH, CURLAUTH_ANY);
+	}
+}
 
-	if(    pData->bulkmode
-	   || (pData->dynSrchIdx == 0 && pData->dynSrchType == 0 && pData->dynParent == 0)) {
-		/* in this case, we know no tpls are involved in the request-->NULL OK! */
-		setCurlURL(pWrkrData, pData, NULL);
+static rsRetVal
+curlSetup(wrkrInstanceData_t *pWrkrData, instanceData *pData)
+{
+	pWrkrData->curlHeader = curl_slist_append(NULL, "Content-Type: text/json; charset=utf-8");
+	pWrkrData->curlPostHandle = curl_easy_init();
+	if (pWrkrData->curlPostHandle == NULL) {
+		return RS_RET_OBJ_CREATION_FAILED;
 	}
+	curlPostSetup(pWrkrData->curlPostHandle, pWrkrData->curlHeader, pData->authBuf);
 
-	if(Debug) {
-		if(pData->dynSrchIdx == 0 && pData->dynSrchType == 0 && pData->dynParent == 0)
-			dbgprintf("omelasticsearch setup, using static REST URL\n");
-		else
-			dbgprintf("omelasticsearch setup, we have a dynamic REST URL\n");
+	pWrkrData->curlCheckConnHandle = curl_easy_init();
+	if (pWrkrData->curlCheckConnHandle == NULL) {
+		curl_easy_cleanup(pWrkrData->curlPostHandle);
+		pWrkrData->curlPostHandle = NULL;
+		return RS_RET_OBJ_CREATION_FAILED;
 	}
+	curlCheckConnSetup(pWrkrData->curlCheckConnHandle, pWrkrData->curlHeader, 
+		pData->healthCheckTimeout, pData->allowUnsignedCerts);
+
 	return RS_RET_OK;
 }
 
 static void
 setInstParamDefaults(instanceData *pData)
 {
-	pData->server = NULL;
-	pData->port = 9200;
+	pData->serverBaseUrls = NULL;
+	pData->defaultPort = 9200;
+	pData->healthCheckTimeout = 3500;
```

Comment:
curl's default is infinite and elastic's default is 30 seconds, both of which seemed way too long if we're trying to process messages quickly and fail fast. 3.5 seconds seems like a reasonable default, that can be extended if users need.